cert-manager/print-your-cert
# Readonly access for the guestbook

Maybe your guestbook server could also serve https://readonly-guestbook.print-your-cert.cert-manager.io/ hopefully with some Let's Encrypt cert instead so that anyone can look at the guestbook from their phone?

I tested with:

```bash
gcloud compute instances create test \                                   
    --project jetstack-mael-valais \
    --network default \
    --machine-type=e2-micro \
    --image-family=debian-12 \
    --image-project=debian-cloud \
    --can-ip-forward \
    --boot-disk-size=10GB \
    --zone=europe-west2-b
IP=$(gcloud compute instances describe test \
       --project jetstack-mael-valais \
       --zone=europe-west2-b --format json \
         | jq -r '.networkInterfaces[].accessConfigs[] | select(.type=="ONE_TO_ONE_NAT") | .natIP')
gcloud dns record-sets update --project cert-manager-io \
     --zone print-your-cert-cert-manager-io \
     --type=A --ttl=300 readonly-guestbook.print-your-cert.cert-manager.io --rrdatas=$IP
```

I don't need to open 80 and 443 as they are already open by default. ~But I had to open 9090~ (no longer needed as I now multiplex TLS connections on 443)

Since guestbook.print-your-cert.cert-manager.io is managed by Tofu (https://github.com/cert-manager/infrastructure/pull/29), I needed to work around that:

```bash
IP=$(gcloud compute instances describe test \
  --project jetstack-mael-valais \
  --zone=europe-west2-b --format json \
  | jq -r '.networkInterfaces[].accessConfigs[] | select(.type=="ONE_TO_ONE_NAT") | .natIP')

sudo perl -ni -e 'print if !/guestbook.print-your-cert.cert-manager.io/' /etc/hosts
sudo tee -a /etc/hosts <<<"$IP guestbook.print-your-cert.cert-manager.io"

ssh pi sudo perl -ni -e '"print if !/guestbook.print-your-cert.cert-manager.io/"' /etc/hosts
ssh pi sudo tee -a /etc/hosts <<<"$IP guestbook.print-your-cert.cert-manager.io"

sudo perl -ni -e 'print if !/readonly-guestbook.print-your-cert.cert-manager.io/' /etc/hosts
sudo tee -a /etc/hosts <<<"$IP readonly-guestbook.print-your-cert.cert-manager.io"

ssh pi sudo perl -ni -e '"print if !/readonly-guestbook.print-your-cert.cert-manager.io/"' /etc/hosts
ssh pi sudo tee -a /etc/hosts <<<"$IP readonly-guestbook.print-your-cert.cert-manager.io"
```

Then, copied the stuff:

```bash
gcloud compute scp --project jetstack-mael-valais --zone=europe-west2-b  \
  ca.crt tls.crt tls.key test:.
gcloud compute ssh --project jetstack-mael-valais --zone=europe-west2-b test -- sudo mkdir -p /etc/guestbook
gcloud compute ssh --project jetstack-mael-valais --zone=europe-west2-b test -- sudo mv ca.crt tls.crt tls.key /etc/guestbook
```

Finally, build and push:

```bash
GOARCH=amd64 GOOS=linux go build -C guestbook .
gcloud compute scp --project jetstack-mael-valais --zone=europe-west2-b  \
  guestbook/guestbook test:.
gcloud compute ssh --project jetstack-mael-valais --zone=europe-west2-b test -- \
  sudo install guestbook /usr/local/bin
```

Then, I started it as a systemd service:

```bash
gcloud compute ssh --project jetstack-mael-valais --zone=europe-west2-b test -- bash <<'EOF'
sudo tee /etc/systemd/system/guestbook.service <<'SVC'
[Service]
RuntimeDirectory=guestbook
ExecStart=guestbook -ca-cert /etc/guestbook/ca.crt -tls-chain /etc/guestbook/tls.crt -tls-key /etc/guestbook/tls.key -listen :443 -domain guestbook.print-your-cert.cert-manager.io -readonly-domain -readonly-listen-insecure :80 readonly-guestbook.print-your-cert.cert-manager.io -prod -db-path /var/run/guestbook/guestbook.sqlite -autocert-dir /var/run/guestbook
ExecStartPre=guestbook -init-db -db-path /var/run/guestbook/guestbook.sqlite
Restart=always
RestartSec=5s

[Install]
WantedBy=multi-user.target
SVC
sudo systemctl daemon-reload
EOF
```

> [!NOTE]
>
> Locally, I used to run:
> ```bash
> guestbook -init-db
> sudo guestbook -ca-cert ca.crt -tls-chain tls.crt -tls-key tls.key \
>   -listen :9090 \
>   -domain guestbook.print-your-cert.cert-manager.io \
>   -readonly-listen-insecure :80 \
>   -readonly-domain readonly-guestbook.print-your-cert.cert-manager.io \
>   -prod
> ```

**Test this "live" test:**

1. Go to https://print-your-cert.cert-manager.io and fill in your email and name.
2. Wait for the cert to be ready and click "Sign the guestbook" (watch out, the real printer is running 😅)
4. Go to https://readonly-guestbook.print-your-cert.cert-manager.io to see the guestbook. To get a "star" instead of a red cross, you need to use `curl` (or any HTTP client) to sign the book.
5. To do that, go back to the page in (1), click "Download cert bundle tar".
6. Open a shell session in the `~/Downloads` folder.
7. Run:
    ```bash
    tar xf cert-manager-bundle.tar
    curl -k https://guestbook.print-your-cert.cert-manager.io/write \
      --cacert ca.crt --cert chain.pem --key pkey.pem \
      -X POST --data-urlencode message@/dev/stdin \
    <<EOF
    Excellent job, Ash! I love this ❤️
    EOF
    ```
8. Go back to https://readonly-guestbook.print-your-cert.cert-manager.io to see the guestbook. Now, you should see a :star:!
